haystack/core/pipeline/base.py

```diff
@@ -1042,24 +1042,134 @@ def _add_missing_input_defaults(component_inputs: Dict[str, Any], component_inpu
 
         return component_inputs
 
-    @staticmethod
+    def _tiebreak_waiting_components(
+        self,
+        component_name: str,
+        priority: ComponentPriority,
+        priority_queue: FIFOPriorityQueue,
+        topological_sort: Union[Dict[str, int], None],
+    ):
+        """
+        Decides which component to run when multiple components are waiting for inputs with the same priority.
+        """
+        components_with_same_priority = [component_name]
+
+        while len(priority_queue) > 0:
+            next_priority, next_component_name = priority_queue.peek()
+            if next_priority == priority:
+                priority_queue.pop()  # actually remove the component
+                components_with_same_priority.append(next_component_name)
+            else:
+                break
+
+        if len(components_with_same_priority) > 1:
+            if topological_sort is None:
+                if networkx.is_directed_acyclic_graph(self.graph):
+                    topological_sort = networkx.lexicographical_topological_sort(self.graph)
+                    topological_sort = {node: idx for idx, node in enumerate(topological_sort)}
+                else:
+                    condensed = networkx.condensation(self.graph)
+                    condensed_sorted = {node: idx for idx, node in enumerate(networkx.topological_sort(condensed))}
+                    topological_sort = {
+                        component_name: condensed_sorted[node]
+                        for component_name, node in condensed.graph["mapping"].items()
+                    }
+
+            components_with_same_priority = sorted(
+                components_with_same_priority, key=lambda comp_name: (topological_sort[comp_name], comp_name.lower())
```

Comment:
We sort the components by topological sort idx first and by component name second. This will tiebreak on topological sort in most cases and fall back to lexicographical tiebreak only when the execution order does not matter (2 components at the same level) or when two waiting components are part of the same cycle.